### ISSUE TITLE
Allow passing variables from calls of training_step when using multiple optimizers

### ISSUE THREAD

**Issue by ifsheldon** [PyTorchLightning/pytorch-lightning#8740]
## 🚀 Feature

Allow passing variables from calls of training_step **when using multiple optimizers**

### Motivation

I am now trying to port the plain PyTorch implementation of Meta Pseudo Label in MPL-pytorch to a Lightning one, so I need to use multiple optimizers. The way the algorithm uses optimizers is very similar to GAN's, so I found and read the [GAN tutorial](https://pytorch-lightning.readthedocs.io/en/latest/notebooks/lightning_examples/basic-gan.html). However, the only difference is that in the second call of `training_step`, I need to have the loss of the first `training_step` call, which seems not possible for now.

### Pitch

With a little bit modification, I'd like to have something like
```python
def training_step(self, batch, batch_idx, optimizer_idx, additional_ctx : Optional[Any] = None):
        # ...

        # train generator
        if optimizer_idx == 0:
            # when optimizer_idx == 0 additional_ctx should be None
            # ...
            output = OrderedDict({'loss': g_loss, 'progress_bar': tqdm_dict, 'log': tqdm_dict})
            additional_ctx = {'stage_1_loss': g_loss}
            return output, additional_ctx

        # train discriminator
        if optimizer_idx == 1:
            stage_1_loss = additional_ctx["stage_1_loss"]
            # ....
            output = OrderedDict({'loss': d_loss, 'progress_bar': tqdm_dict, 'log': tqdm_dict})
            return output, None
```

### Alternatives

I am thinking about if we can use decorators to make it cleaner. Something like this seems great.
```python
@pl.training_step(opt_idx=0, on_step_end="train_generator_end")
def train_generator(self, batch, batch_idx,  additional_ctx : Optional[Any] = None):
     # ...
     output = OrderedDict({'loss': g_loss, 'progress_bar': tqdm_dict, 'log': tqdm_dict})
     additional_ctx = {'stage_1_loss': g_loss}
     return output, additional_ctx

@pl.training_step(opt_idx=1, on_step_end="train_discriminator_end")
def train_discriminator(self, batch, batch_idx,  additional_ctx : Optional[Any] = None):
     stage_1_loss = additional_ctx["stage_1_loss"]
     # ....
     output = OrderedDict({'loss': d_loss, 'progress_bar': tqdm_dict, 'log': tqdm_dict})
     return output, None

def train_generator_end(...):
     ....
     pass

def train_discriminator_end(...):
     ....
```

### Additional context

There're some issues to be considered:
* whether to allow passing tensors that are not `detach`ed
* the decorator one seems to be another improvement that is a bit parallel

**Comment by ananthsub** [comment]
For use of multiple optimizers, is manual optimization something that would make porting the project from vanilla PyTorch to Lightning easier?

**Comment by ifsheldon** [comment]
Well, yes and no. With auto optimization, we can keep other features available. As I discussed with tchaton on Slack, we cannot have auto gradient accumulation for example when we enable manual optimization.

**Comment by ananthsub** [comment]
> we may need arbitrary tensors for later use.

the arbitrary tensors come from within the training step though, right?

**Comment by ifsheldon** [comment]
I mean the general case will be, in the n-th `training_step` call, we may need some tensors calculated in previous (n-1) `training_step` calls.

**Comment by ananthsub** [comment]
> we may need some tensors calculated in previous (n-1) training_step calls.

In this case, you can store these tensors on the LightningModule to be used across calls

**Comment by ifsheldon** [comment]
I think it's definitely okay with non-distributed training. But I wonder if it has some issue when it comes with distributed training.

**Comment by ananthsub** [comment]
> I think it's definitely okay with non-distributed training. But I wonder if it has some issue when it comes with distributed training.

In this case, the Lightning trainer still doesn't all-gather tensors across ranks in distributed training. We can offer utilities to make this easier which you could call in your LightningModule to support distributed training. https://github.com/PyTorchLightning/pytorch-lightning/issues/7534 aims to help with this

**Comment by ifsheldon** [comment]
OK, then `detach`ed tensors seem good! But how about passing un`detach`ed tensors?

**Comment by ananthsub** [comment]
> OK, then `detach`ed tensors seem good! But how about passing un`detach`ed tensors?

n00b question: How would you usually express this in a vanilla PyTorch training loop? We can try to follow that as closely as possible with lightning here

**Comment by ifsheldon** [comment]
I mean what if we want to retain a previous compute graph for a later `training_step` call. But I think it's rather complex in this case, it's time to go to manual optimization.

Nonetheless, using decorators seems good to me? Any opinions?
